### PR TITLE
64 implement parse command

### LINKED ISSUE #64: Implement `parse_command`
## TODO

- [x] Update `grammer.md`
- [x] Implement `parse_command`
- [x] Test `parse_command`
- [x] Replace `handle_error` with `handle_syntax_error`

### grammer

```bison
 pipeline         :              command
                  | pipeline '|' command
                  ;
 command          : simple_command
                  | '(' list ')'
                  | '(' list ')' redirect_list
                  ;
```

### LINKED ISSUE #66: Replace `newline` with `EOF`
### fix error message
https://github.com/rask24/minishell/blob/b7d538e9916dc5580b3351182866e805ef34dd04/src/parser/parser_utils.c#L35

### PR DESCRIPTION
**Pull request by rask24**
fix #64
fix #66 

- **Update grammer.md**
- **Refactor parse_simple_command**
- **Implement parse_command.c**

## Summary by CodeRabbit

- **New Features**
	- Enhanced command parsing capabilities, including support for nested commands and redirection.
	- Introduced new parsing functions for commands and redirection handling.

- **Bug Fixes**
	- Improved error handling in command parsing to allow for more robust processing of unexpected node types.
	- Updated error messages for better clarity regarding syntax errors.

- **Tests**
	- Added a comprehensive suite of unit tests for the new command parsing functionality, covering various scenarios and edge cases.
	- Expanded test coverage for parsing lists and pipelines, including validation of subshells and handling of invalid tokens.

### CODE REVIEW

File: src/parser/parse_command.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		return (parse_simple_command(cur_token));
+	consume_token(cur_token);
+	node = parse_list(cur_token);
+	if (node == NULL)
+		return (NULL);
+	if (!expect_token(cur_token, TOKEN_R_PARENTHESIS))
+		return (handle_error(node, get_token_value(*cur_token)));
+	if (is_redirect_token(get_token_type(*cur_token)))
+	{
+		if (!parse_redirect(node, cur_token))
```
Comment (alex216) [inline src/parser/parse_command.c:37]: since `parse_redirect()` is called here and other than from `parse_simple_command()`, we must check syntax error inside `parse_redirect()` instead of in front of `parse_simple_command()`.

File: src/parser/parse_command.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+	if (get_token_type(*cur_token) != TOKEN_L_PARENTHESIS)
+		return (parse_simple_command(cur_token));
+	consume_token(cur_token);
+	node = parse_list(cur_token);
+	if (node == NULL)
+		return (NULL);
+	if (!expect_token(cur_token, TOKEN_R_PARENTHESIS))
+		return (handle_error(node, get_token_value(*cur_token)));
+	if (is_redirect_token(get_token_type(*cur_token)))
```
Comment (alex216) [inline src/parser/parse_command.c:35]: maybe this if condition should check `EOF` instead of `redirect`

**Review by alex216 — COMMENTED** [review]
@rask24 please check 2 comments

File: src/parser/parse_command.c (hunk 1 above)
Comment (alex216) [reply src/parser/parse_command.c:39]: this test will fail.
```
TEST(parse_command, InvalidTokenRedirect1) {
  t_token_list *token_list = nullptr;

  // (ls -l ) |
  ft_lstadd_back(&token_list,
                 construct_token(TOKEN_L_PARENTHESIS, strdup("(")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_WORD, strdup("ls")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_WORD, strdup("-l")));
  ft_lstadd_back(&token_list,
                 construct_token(TOKEN_R_PARENTHESIS, strdup(")")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_PIPE, strdup("|")));
  ft_lstadd_back(&token_list, construct_token(TOKEN_EOF, nullptr));

  t_ast *node = parse_command(&token_list);

  EXPECT_EQ(node, nullptr);

  destroy_token_list(token_list);
}
```

**Review by alex216 — COMMENTED** [review]
@rask24  please check 1 comment above.

File: src/parser/parse_command.c (hunk 1 above)
Comment (rask24) [reply src/parser/parse_command.c:39]: @alex216 
I think this test is not appropriate for `parse_command` but suitable for `parse_list` or `parse_pipeline`. 
`parse_command` is designed to handle only simple_command, subshell, and redirects.

To improve our test coverage, we need to add new tests specifically for parse_pipeline and parse_list to cover more complex command structures.

Comment (rask24) [reply src/parser/parse_command.c:39]: To improve our test coverage, we need to add new tests for parse_pipeline and parse_list to cover more complex command structures.

Comment (alex216) [reply src/parser/parse_command.c:39]: I was wrong. you are right, the test should be for `parse_pipeline()` and it works perfectly already!
